python/rockets/client.py

```diff
@@ -0,0 +1,136 @@
+        return self._client.url()
+
+    @copydoc(AsyncClient.connected)
+    def connected(self):  # noqa: D102 pylint: disable=missing-docstring
+        return self._client.connected()
+
+    @copydoc(AsyncClient.connect)
+    def connect(self):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.connect())
+
+    @copydoc(AsyncClient.disconnect)
+    def disconnect(self):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.disconnect())
+
+    @copydoc(AsyncClient.as_observable)
+    def as_observable(self):  # noqa: D102 pylint: disable=missing-docstring
+        return self._client.as_observable()
+
+    @copydoc(AsyncClient.send)
+    def send(self, message):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.send(message))
+
+    @copydoc(AsyncClient.notify)
+    def notify(self, method, params=None):  # noqa: D102 pylint: disable=missing-docstring
+        self._verify_environment()
+        self._call_sync(self._client.notify(method, params))
+
+    @copydoc(AsyncClient.request)
+    def request(self, method, params=None, response_timeout=None):  # noqa: D102,D205 pylint: disable=C0111,W9011,W9012,W9015,W9016
+        """
+        :param int response_timeout: number of seconds to wait for the response
+        :raises TimeoutError: if request was not answered within given response_timeout
+        """
+        self._verify_environment()
+        return self._call_sync(self._client.request(method, params), response_timeout)
+
+    @copydoc(AsyncClient.batch)
+    def batch(self, requests, response_timeout=None):  # noqa: D102,D205 pylint: disable=C0111,W9011,W9012,W9015,W9016
+        """
+        :param int response_timeout: number of seconds to wait for the response
+        :raises TimeoutError: if request was not answered within given response_timeout
+        """
+        self._verify_environment()
+        return self._call_sync(self._client.batch(requests), response_timeout)
+
+    @copydoc(AsyncClient.async_request)
+    def async_request(self, method, params=None):  # noqa: D102 pylint: disable=missing-docstring
+        return self._async_client.async_request(method, params)
+
+    @copydoc(AsyncClient.batch)
+    def async_batch(self, requests):  # noqa: D102 pylint: disable=missing-docstring
+        return self._async_client.async_batch(requests)
+
+    def _verify_environment(self):
+        if not self._thread and self._client.loop().is_running():
+            raise RuntimeError("Unknown working environment")
+
+    def _call_sync(self, original_function, response_timeout=None):
+        if self._thread:
+            future = asyncio.run_coroutine_threadsafe(
+                original_function,
+                self._client.loop()
+            )
+            return future.result(response_timeout)
```

Comment:
I'm new to this so it may be a silly question, but I'm struggling to understand this whole execution of synchronous calls inside a new thread... If the loop is already running, couldn't the above simply be:
```
task = self._client.loop().ensure_future(original_function)
return task.result()
```
?

Reply:
`ensure_future` just schedules it for execution. `result()` is not blocking, hence you get an invalid state if it is not ready yet. And as the loop is already running and currently execute your current statements/functions, it cannot execute something else in parallel.

Reply:
my mistake, future.result() indeed does not block; what I meant then is to `await task` before that. Are you saying this blocks the processing loop entirely? I imagined it would just yield control to the coroutine(s) doing the work and resume here when the result is ready (without the need to call ` loop.run_until_complete(future)` since the loop is already running).

Reply:
That's what you would intuitively expect to happen, yes :smile: But `await` can only be called from within an `async` function, hence all the hassle.

Reply:
errr, but if the loop is running, how can this code not be called from inside an async function already..?

Reply:
Huh! I'm pretty sure I was doing this at one point as this is what I would have expected it to work from the beginning. Testing it now, it seems it does work, so probably this threading madness is not needed at all! Thx for insisting on that :smile: I continue verifying that, but it looks good.

Reply:
Works technically, but the user must then do `await` or `run_until_complete()` himself if he wants blocking calls. I can live with that, not sure other users might.

Reply:
OK, isn't it possible to wrap this behavior entirely inside this class? I don't see exactly what you mean, maybe we could have a look at the new code in person.

Reply:
Ok then, I will check in person :smile:

Reply:
Resolved now by only providing a synchronous wrapper on the AsyncClient, but no second client for async access. That requires creating an AsyncClient in the first place.